crypto/ec/ecp_nistp384.c

```diff
@@ -0,0 +1,1931 @@
+{
+    memset(out, 0, 56);
+    out[0] = (*((limb *) & in[0])) & bottom56bits;
+    out[1] = (*((limb_aX *) & in[7])) & bottom56bits;
+    out[2] = (*((limb_aX *) & in[14])) & bottom56bits;
+    out[3] = (*((limb_aX *) & in[21])) & bottom56bits;
+    out[4] = (*((limb_aX *) & in[28])) & bottom56bits;
+    out[5] = (*((limb_aX *) & in[35])) & bottom56bits;
+    memmove(&out[6], &in[42], 6);
+}
+
+static void felem_to_bin48(u8 out[48], const felem in)
+{
+    memset(out, 0, 48);
+    (*((limb *) & out[0]))     |= (in[0] & bottom56bits);
+    (*((limb_aX *) & out[7]))  |= (in[1] & bottom56bits);
+    (*((limb_aX *) & out[14])) |= (in[2] & bottom56bits);
+    (*((limb_aX *) & out[21])) |= (in[3] & bottom56bits);
+    (*((limb_aX *) & out[28])) |= (in[4] & bottom56bits);
+    (*((limb_aX *) & out[35])) |= (in[5] & bottom56bits);
+    memmove(&out[42], &in[6], 6);
+}
+
+/* BN_to_felem converts an OpenSSL BIGNUM into an felem */
+static int BN_to_felem(felem out, const BIGNUM *bn)
+{
+    felem_bytearray b_out;
+    int num_bytes;
+
+    if (BN_is_negative(bn)) {
+        ERR_raise(ERR_LIB_EC, EC_R_BIGNUM_OUT_OF_RANGE);
+        return 0;
+    }
+    num_bytes = BN_bn2lebinpad(bn, b_out, sizeof(b_out));
+    if (num_bytes < 0) {
+        ERR_raise(ERR_LIB_EC, EC_R_BIGNUM_OUT_OF_RANGE);
+        return 0;
+    }
+    bin48_to_felem(out, b_out);
+    return 1;
+}
+
+/* felem_to_BN converts an felem into an OpenSSL BIGNUM */
+static BIGNUM *felem_to_BN(BIGNUM *out, const felem in)
+{
+    felem_bytearray b_out;
+    felem_to_bin48(b_out, in);
+    return BN_lebin2bn(b_out, sizeof(b_out), out);
+}
+
+/*-
+ * Field operations
+ * ----------------
+ */
+
+static void felem_one(felem out)
+{
+    out[0] = 1;
+    memset(&out[1], 0, sizeof(limb) * (NLIMBS-1));
+}
+
+static void felem_assign(felem out, const felem in)
+{
+    memcpy(out, in, sizeof(felem));
+}
+
+/* felem_sum64 sets out = out + in. */
+static void felem_sum64(felem out, const felem in)
+{
+    unsigned int i;
+
+    for (i = 0; i < NLIMBS; i++)
+        out[i] += in[i];
+}
+
+/* felem_scalar sets out = in * scalar */
+static void felem_scalar(felem out, const felem in, limb scalar)
+{
+    unsigned int i;
+
+    for (i = 0; i < NLIMBS; i++)
+        out[i] = in[i] * scalar;
+}
+
+/* felem_scalar64 sets out = out * scalar */
+static void felem_scalar64(felem out, limb scalar)
+{
+    unsigned int i;
+
+    for (i = 0; i < NLIMBS; i++)
+        out[i] *= scalar;
+}
+
+/* felem_scalar128 sets out = out * scalar */
+static void felem_scalar128(widefelem out, limb scalar)
+{
+    unsigned int i;
+
+    for (i = 0; i < 2*NLIMBS-1; i++)
+        out[i] *= scalar;
+}
+
+/*-
+ * felem_neg sets |out| to |-in|
+ * On entry:
+ *   in[i] < 2^60 - 2^29
+ * On exit:
+ *   out[i] < 2^60
+ */
+static void felem_neg(felem out, const felem in)
+{
+    /* In order to prevent underflow, we subtract from a multiple of p. */
+    static const limb two60m52m4 = (((limb) 1) << 60)
```

Comment:
I am still struggling to get my head around how you come up with these static constants.. Can you give some insight into how you come up with the numbers below.. What is the actual multiple of p here? (x % p = 0 should be zero)..
I can see that the range seems top be 56 bits with the 2^60-2^4 part (with zeros in the top and bottom bits).

Reply:
In `felem_neg` it's just $2^{12} p_{384}$. See $2^{384} \cdot 2^{12} = 2^{60} \cdot (2^{56})^6$. We then use telescopic sums to introduce a $2^{60}$ term in every limb.

In each routine I use a simple power of two to get our multiple. Hope that makes sense.

Reply:
I may as well do this one in full:
$2^{12} p_{384} = 2^{396} - 2^{140} - 2^{108} + 2^{44} - 2^{12} = 2^{60} (2^{56})^6 - 2^{28} (2^{56})^2 - 2^{52} \cdot 2^{56} + 2^{44} - 2^{12}$
Now introducing into limbs 0 .. 5 a $2^60$ term:
$= (2^{60} - 2^4) (2^{56})^6 + (2^{60} - 2^4) (2^{56})^5 + (2^{60} - 2^4) (2^{56})^4 + (2^{60} - 2^4) (2^{56})^3 + (2^{60} - 2^{28} - 2^4) (2^{56})^2 + (2^{60} - 2^{52} - 2^4) 2^{56} + (2^{60} + 2^{44} - 2^{12})$

Reply:
Thanks for explaining (and your patience).. That makes sense..
Is felem_diff_128_64 similiar?
Would you maybe be able to add comments explaining this in multiple functions, so others can also understand what is going on.

Reply:
Not a problem. Yep I was basically just copying the style of ecp_nistp{224,521}.c, but it doesn't hurt to just state the multiple up front.

Precisely, in `felem_diff_128_64` it's $2^{16}p_{384}$, and in `felem_reduce` it's $2^{76}p_{384}$.

Reply:
Is the felem_diff_128_64 still 2^56 based

Reply:
Yes, so irrespective of the machine width - `widefelem` or `felem` and irrespective of the number ($k$) of limbs, the value being represented is always:

$\sigma^{k}_{i=0} (2^{56})^i l_i$